### PR TITLE
Introduce dereference operator for references

### LINKED ISSUE #2803: Need a way get a dereferenced copy from a reference
### Issue to be solved

Sometimes it is useful to be able to get a dereferenced-copy from a reference.

For e.g. assume a large dictionary is in storage, and one would need to get a copy of just one entry of that dictionary. In the latest released version, this is possible by:
- First, borrowing a reference to the dictionary (say `dictionaryRef`)
- Accessing the entry using index-access `dictionaryRef[key]`, which would return a copy of the value at the `key`.

However, with stable cadence, given the index syntax `dictionaryRef[key]` also returns another reference, the above approach no longer works. Instead, they would need to 
- Take a "copy" of the entire dictionary first (say `dictionaryCopy`)
- Then, get a copy of the value using the index-access `dictionaryCopy[key]`.
The downside of this is now they would need to copy the entire outer dictionary which could be a very expensive operation.

### Suggested Solution

It would be nice to have a way to get a dereferenced copy from a reference (say a built-in method `dereference()`), so that they can get a copy from the reference itself.
i.e:
```cadence
let elementRef = dictionaryRef[key]

let elementCopy = dereference(elementRef)
```

Some suggestions for the name of the built-in method:
- `dereference()`
- `dereferenceCopy()`
- Could also use the c-like dereference operator- asterisk (`*`), but I personally prefer one of the above, because they are verbose.

### PR DESCRIPTION
**Pull request by darkdrag00nv2**
Closes #2803

## Description

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Comment by darkdrag00nv2** [comment]
Need to cover a few more types in the interpreter tests.

File: runtime/sema/type.go [hunk 1]
```diff
@@ -6309,6 +6307,78 @@ func (t *ReferenceType) Resolve(typeArguments *TypeParameterTypeOrderedMap) Type
...
 
+func (t *ReferenceType) GetMembers() map[string]MemberResolver {
+	t.initializeMemberResolvers()
+	return t.memberResolvers
+}
+
+const ReferenceTypeDereferenceFunctionName = "dereference"
+
+const referenceTypeDereferenceFunctionDocString = `
+	Returns a copy of the reference value after dereferencing.
```
Comment (dsainati1) [inline runtime/sema/type.go:6318]:
```suggestion
	Returns a copy of the referenced value after dereferencing.
```

Comment (dsainati1) [reply runtime/sema/type.go:6318]: IMO we should also add something to doc string about how this only works for primitive values

File: runtime/sema/type.go [hunk 2]
```diff
@@ -6448,6 +6518,33 @@ func (t *AddressType) initializeMemberResolvers() {
 	})
 }
 
+func IsPrimitiveOrContainerOfPrimitive(ty Type) bool {
+	if ty.IsPrimitiveType() {
+		return true
+	}
+
+	// TODO: Do we also want to count Dictionary?
```
Comment (dsainati1) [inline runtime/sema/type.go:6526]: Yes! Dictionaries should also be considered containers the same way arrays are

File: runtime/sema/type.go [hunk 3]
```diff
@@ -6448,6 +6518,33 @@ func (t *AddressType) initializeMemberResolvers() {
...
+	switch ty.(type) {
```
Comment (dsainati1) [inline runtime/sema/type.go:6527]:
```suggestion
	switch typedTy := ty.(type) {
```

Comment (dsainati1) [reply runtime/sema/type.go:6527]: This way you don't have to manually cast inside the `case`s below, or panic on an unreachable failure.

File: runtime/tests/interpreter/reference_test.go [hunk 4]
```diff
@@ -1815,3 +1816,1073 @@ func TestInterpretReferenceToReference(t *testing.T) {
 		require.ErrorAs(t, err, &interpreter.NestedReferenceError{})
 	})
 }
+
+func TestInterpretReferenceDereference(t *testing.T) {
```
Comment (dsainati1) [inline runtime/tests/interpreter/reference_test.go:1820]: Can you add a test for optional references using optional chaining? I.e. something of type `&Int?` being dereferenced using `ref?.dereference()`?

File: runtime/sema/type.go (hunk 2 above)
Comment (darkdrag00nv2) [reply runtime/sema/type.go:6526]: So the main question is, do we want `Dictionary<PrimitiveKey, Non-PrimitiveValue>` or `Dictionary<Non-Primitive key, Primitive value` to also support `dereference`?

I guess we only want `Dictionary<primitive key, primitive value>` to support dereference but wanted to confirm.

Comment (dsainati1) [reply runtime/sema/type.go:6526]: Dictionary keys are already restricted to be subtypes of "Hashable Structs" (see `IsHashableStructType`), which similar to the Primitive type designation (and might actually be able to be refactored to use the new `IsPrimitiveType` method). Point is though, that I don't think you need to worry about the key types at all when checking whether a dictionary reference is dereference-able, just check the values.

File: runtime/sema/type.go (hunk 3 above)
Comment (turbolent) [reply runtime/sema/type.go:6527]: Or even shadow: `ty := ty.(type)`

File: runtime/sema/errors.go [hunk 5]
```diff
@@ -2755,6 +2755,29 @@ func (e *InvalidResourceOptionalMemberError) Error() string {
...
+var _ SemanticError = &InvalidMemberError{}
+var _ errors.UserError = &InvalidMemberError{}
+
+func (*InvalidMemberError) isSemanticError() {}
+
+func (*InvalidMemberError) IsUserError() {}
+
+func (e *InvalidMemberError) Error() string {
+	return fmt.Sprintf(
+		"%s `%s` is not available for the type",
```
Comment (turbolent) [inline runtime/sema/errors.go:2775]: Maybe add a reason here, i.e. add a `Reason string` field to the error type, and maybe include it if non-empty with `": %s"`

File: runtime/interpreter/value.go [hunk 6]
```diff
@@ -19855,6 +19855,24 @@ func (v *StorageReferenceValue) GetMember(
...
+				return self.Transfer(
+					invocation.Interpreter,
+					invocation.LocationRange,
+					atree.Address{},
+					false,
+					nil,
+					nil,
+				)
+			},
+		)
```
Comment (turbolent) [inline runtime/interpreter/value.go:19873]: Maybe refactor this into a function and use it here and in `EphemeralReferenceValue`, so it is ensured that the same code is used in both values

**Review by turbolent — COMMENTED** [review]
Nice!

**Comment by turbolent** [comment]
CI is reporting data races: https://github.com/onflow/cadence/actions/runs/7240009930/job/19722511487?pr=2982

File: runtime/sema/type.go (hunk 3 above)
Comment (darkdrag00nv2) [reply runtime/sema/type.go:6527]: thanks, done.

File: runtime/sema/type.go (hunk 2 above)
Comment (darkdrag00nv2) [reply runtime/sema/type.go]: Done

File: runtime/sema/errors.go (hunk 5 above)
Comment (darkdrag00nv2) [reply runtime/sema/errors.go]: Done

File: runtime/interpreter/value.go (hunk 6 above)
Comment (darkdrag00nv2) [reply runtime/interpreter/value.go]: Done

File: runtime/tests/interpreter/reference_test.go (hunk 4 above)
Comment (darkdrag00nv2) [reply runtime/tests/interpreter/reference_test.go:1820]: Done in 43d998901

File: runtime/sema/type.go [hunk 7]
```diff
@@ -6309,6 +6307,84 @@ func (t *ReferenceType) Resolve(typeArguments *TypeParameterTypeOrderedMap) Type
...
+							ReferenceDereferenceFunctionType(t.Type),
+							referenceTypeDereferenceFunctionDocString,
+						)
+					},
+				},
+			},
+		}
+
+		for _, member := range members {
+			resolvers[member.name] = member.resolver
```
Comment (turbolent) [inline runtime/sema/type.go:6374]: If there is already a member resolve, it will be overwritten here

**Review by turbolent — COMMENTED** [review]
Looks good, great work!

In general I'm still not sure if this should be a magic function on references, or rather something else, like an operator.

For the functionality to be a magic function, the PR should also handle the case where a composite declares a member named the same as the magic function, as it seems like the current implementation shadows / makes it inaccessible. We might want to forbid declaring members which have that magic name. An operator would not have that problem.

**Comment by turbolent** [comment]
We might want to use an operator here instead of / in addition to the method, e.g. the unary prefix operator `*`, like in Rust, Go, C, C++, etc.

**Comment by turbolent** [comment]
@darkdrag00nv2 Not sure if you started yet, but if not: I'm half-way done and could finish and push it up here

**Comment by darkdrag00nv2** [comment]
@turbolent Haven't started yet. You can push it. Thank you!

**Comment by turbolent** [comment]
Refactored the `dereference` function to the prefix `*` operator.

@onflow/cadence could you please have a look?

**Review by dsainati1 — COMMENTED** [review]
Now that `dereference` isn't a member function, it would be nice for `*` to have simple/clean behavior with respect to optional references. I.e. if we have a `&Int?` type, it would be nice for the `*` operator on that type to produce a `Int?`. Previously we could do this with `?.` chaining, but with the change to an operator we lose this functionality. Can we add this (and tests for it)?

**Comment by turbolent** [comment]
@dsainati1 Would this simply be `*(optRef!)` ?

**Comment by dsainati1** [comment]
Yeah, it would be nice for the `*` operator to implicitly "lift" itself to operate on optional values as well as normal ones.

**Comment by turbolent** [comment]
@dsainati1 Sounds good! Let's get that added in a follow-up PR

File: runtime/parser/expression_test.go
Comment (SupunS) [inline runtime/parser/expression_test.go:1]: Can you maybe add a test for an expression like `a**b` to ensure it parses correctly / presidency is correct?
i.e: should result in `(a*(*b))`

File: runtime/sema/type.go [hunk 8]
```diff
@@ -6448,6 +6448,25 @@ func (t *AddressType) initializeMemberResolvers() {
...
+		return IsPrimitiveOrContainerOfPrimitive(ty.Type)
+
+	case *ConstantSizedType:
+		return IsPrimitiveOrContainerOfPrimitive(ty.Type)
+
+	case *DictionaryType:
+		return IsPrimitiveOrContainerOfPrimitive(ty.ValueType)
+	}
+
+	return false
```
Comment (SupunS) [inline runtime/sema/type.go:6467]: Just a nit-pick: calling for `IsPrimitiveType` for arrays and dictionaries is guranteed to return false.

```suggestion
	switch ty := ty.(type) {
	case *VariableSizedType:
		return IsPrimitiveOrContainerOfPrimitive(ty.Type)

	case *ConstantSizedType:
		return IsPrimitiveOrContainerOfPrimitive(ty.Type)

	case *DictionaryType:
		return IsPrimitiveOrContainerOfPrimitive(ty.ValueType)
		
	default:
	  	return ty.IsPrimitiveType()
	}
```

File: runtime/parser/expression_test.go
Comment (turbolent) [reply runtime/parser/expression_test.go:1]: Good idea, added in 01ea3f3860478c32eea8b09d3b0f60eafc3de1a0

**Review by SupunS — APPROVED** [review]
LGTM!

**Review by dsainati1 — APPROVED** [review]

**Pull request merged**